tensorflow/core/kernels/adjust_hue_op_gpu.cu.cc

```diff
@@ -0,0 +1,147 @@
+  const float new_h = h * 6.0f;
+  const float chroma = v * s;
+  const float x = chroma * (1.0 - fabsf(fmodf(new_h, 2.0f) - 1.0f));
+  const float new_m = v - chroma;
+  const bool between_0_and_1 = new_h >= 0.0 && new_h < 1;
+  const bool between_1_and_2 = new_h >= 1.0 && new_h < 2;
+  const bool between_2_and_3 = new_h >= 2 && new_h < 3;
+  const bool between_3_and_4 = new_h >= 3 && new_h < 4;
+  const bool between_4_and_5 = new_h >= 4 && new_h < 5;
+  const bool between_5_and_6 = new_h >= 5 && new_h < 6;
+  tuple.r = chroma * (between_0_and_1 || between_5_and_6) +
+    x * (between_1_and_2 || between_4_and_5) + new_m;
+  tuple.g = chroma * (between_1_and_2 || between_2_and_3) +
+    x * (between_0_and_1 || between_3_and_4) + new_m;
+  tuple.b =  chroma * (between_3_and_4 || between_4_and_5) +
+    x * (between_2_and_3 || between_5_and_6) + new_m;
+  return tuple;
+}
+
+__global__ void adjust_hue_nhwc(const int number_elements,
+                                const float * const input, float * const output, const float * const hue_delta) {
+  const float delta = hue_delta[0];
+  // multiply by 3 since we're dealing with contiguous RGB bytes for each pixel (NHWC)
+  const int idx = (blockDim.x * blockIdx.x + threadIdx.x) * 3;
+  // bounds check
+  if (idx > number_elements) {
+    return;
+  }
+  const HsvTuple hsv = rgb2hsv_cuda(input[idx], input[idx+1], input[idx+2]);
+  // hue adjustment
+  float new_h = fmodf(hsv.h + delta, 1.0f);
+  if (new_h < 0.0f) {
+    new_h = fmodf(1.0f + new_h, 1.0f);
+  }
+  const RgbTuple rgb = hsv2rgb_cuda(new_h, hsv.s, hsv.v);
+  output[idx] = rgb.r;
+  output[idx + 1] = rgb.g;
+  output[idx + 2] = rgb.b;
+}
+} // namespace internal
+
+template <>
+class AdjustHueOp<GPUDevice> : public AdjustHueOpBase {
+public:
+explicit AdjustHueOp(OpKernelConstruction* const context)
+  : AdjustHueOpBase(context) {}
+
+void DoCompute(OpKernelContext* const context,
+  const ComputeOptions& options) override {
+  const Tensor* input = options.input;
+  const Tensor* delta = options.delta;
+  Tensor* const output = options.output;
+  const int64 number_elements = input->NumElements();
+  const GPUDevice &d = context->eigen_gpu_device();
+  const auto stream = d.stream();
+  OP_REQUIRES(context, stream, errors::Internal("No GPU stream available."));
+  if (number_elements > 0) {
+    const CudaLaunchConfig config = GetCudaLaunchConfig(number_elements, d);
+    const float * const input_data = input->flat<float>().data();
+    const float * const delta_h = delta->flat<float>().data();
+    float * const output_data = output->flat<float>().data();
+    const int threads_per_block = config.thread_per_block;
+    const int block_count = (number_elements + threads_per_block - 1) / threads_per_block;
+    internal::adjust_hue_nhwc<<<block_count, threads_per_block, 0, stream>>>(
+      number_elements, input_data, output_data, delta_h
+    );
+  }
+}
+};
+
+REGISTER_KERNEL_BUILDER(Name("AdjustHue").Device(DEVICE_GPU),
```

Comment:
Ah! I am a bit surprised that this worked at all. 

Older nvcc cannot handle OpKernel and its registration process. So we had to do all the OpKernel in the CPU translation unit, and only call out an launch function into the Cuda translation unit. 

It might be safer to follow that convention for the multiple Cuda compilers we are supporting. Here is an example, there are many in the kernels directory. 

1. GPU kernel registration in the CPU translation unit: 
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/kernels/segment_reduction_ops.cc#L315
2. The kernel does all the OpKernel related things in this file, and call out to its GPU functor at. Only simpler data structure crosses the file boundary. 
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/kernels/segment_reduction_ops.cc#L291
3. The GPU functor is defined here: 
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/kernels/segment_reduction_ops_gpu.cu.cc#L59
4. And instantiated here: 
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/kernels/segment_reduction_ops_gpu.cu.cc#L95